### PR TITLE
Optional canvas functionality

### PR DESCRIPTION
**Pull request by osc2nuke**
Maybe it can help others.

### CODE REVIEW

File: lib/ImageResize.php [hunk 1]
```diff
@@ -198,22 +198,55 @@ public function imageCreateJpegfromExif($filename)
         return $img;
     }
 
+    /**
+     * Create canvas
+     *
+     * recreate width
+	 * recreate height
```
Comment (peter279k) [inline lib/ImageResize.php]: Can you organize this code?
The space is not very well.

File: lib/ImageResize.php [hunk 2]
```diff
@@ -198,22 +198,55 @@ public function imageCreateJpegfromExif($filename)
...
+	 * center image
+    */
+    function drawCanvas()
+    {
+		// set new width and height for image
```
Comment (peter279k) [inline lib/ImageResize.php]: It's same as the previous comment.

File: lib/ImageResize.php [hunk 3]
```diff
@@ -198,22 +198,55 @@ public function imageCreateJpegfromExif($filename)
...
+      $ratio = ($ratiow > $ratioh) ? $this->getDestWidth() / $this->original_w : $this->getDestHeight() / $this->original_h;
+      if($this->original_w > $this->getDestWidth() || $this->original_h > $this->getDestHeight()) { 
+         $this->new_w = round($this->original_w * $ratio); 
+         $this->new_h = round($this->original_h * $ratio); 
+     } else {
+         $this->new_w = round($this->original_w); 
+         $this->new_h = round($this->original_h);
+     }
+     
+		// center position image
```
Comment (peter279k) [inline lib/ImageResize.php]: It's same as the previous comment.

File: lib/ImageResize.php [hunk 4]
```diff
@@ -198,22 +198,55 @@ public function imageCreateJpegfromExif($filename)
...
      *
      * @param string $filename
+     * @param boolean $draw_canvas
      * @param string $image_type
      * @param integer $quality
      * @param integer $permissions
      * @return \static
-     */
-    public function save($filename, $image_type = null, $quality = null, $permissions = null)
+     */
```
Comment (peter279k) [inline lib/ImageResize.php]: Is this the additional space?

File: lib/ImageResize.php [hunk 5]
```diff
@@ -198,22 +198,55 @@ public function imageCreateJpegfromExif($filename)
...
-    public function save($filename, $image_type = null, $quality = null, $permissions = null)
+     */	
+    function save($filename, $draw_canvas = false, $image_type = null, $quality = null, $permissions = null)
     {
         $image_type = $image_type ?: $this->source_type;
         $quality = is_numeric($quality) ? (int) abs($quality) : null;
 
         switch ($image_type) {
-        case IMAGETYPE_GIF:
+            case IMAGETYPE_GIF:
```
Comment (peter279k) [inline lib/ImageResize.php]: The space is not very well.
Please fix this.

File: lib/ImageResize.php [hunk 6]
```diff
@@ -222,14 +255,14 @@ public function save($filename, $image_type = null, $quality = null, $permission
             imagesavealpha($dest_image, true);
             break;
 
-        case IMAGETYPE_JPEG:
+            case IMAGETYPE_JPEG:
```
Comment (peter279k) [inline lib/ImageResize.php]: It's the same as previous comment.

File: lib/ImageResize.php [hunk 7]
```diff
@@ -222,14 +255,14 @@ public function save($filename, $image_type = null, $quality = null, $permission
...
             $dest_image = imagecreatetruecolor($this->getDestWidth(), $this->getDestHeight());
 
             $background = imagecolorallocate($dest_image, 255, 255, 255);
             imagefilledrectangle($dest_image, 0, 0, $this->getDestWidth(), $this->getDestHeight(), $background);
             break;
 
-        case IMAGETYPE_WEBP:
+            case IMAGETYPE_WEBP:
```
Comment (peter279k) [inline lib/ImageResize.php]: It's the same as previous comment.

File: lib/ImageResize.php [hunk 8]
```diff
@@ -239,7 +272,7 @@ public function save($filename, $image_type = null, $quality = null, $permission
             imagefilledrectangle($dest_image, 0, 0, $this->getDestWidth(), $this->getDestHeight(), $background);
             break;
 
-        case IMAGETYPE_PNG:
+            case IMAGETYPE_PNG:
```
Comment (peter279k) [inline lib/ImageResize.php]: It's the same as previous comment.

File: lib/ImageResize.php [hunk 9]
```diff
@@ -254,67 +287,82 @@ public function save($filename, $image_type = null, $quality = null, $permission
             imagesavealpha($dest_image, true);
             break;
         }
-
+
```
Comment (peter279k) [inline lib/ImageResize.php]: It looks like the additional space.

**Comment by peter279k** [comment]
Can you explain why some people require the ```draw_canvas``` functionality?

**Comment by osc2nuke** [comment]
I have no idea why i always have problems with indenting/spacing's on my codes.
[personal experiences]
Reasons for a canvas is that especially shop owners encounter dis-formed layouts when they using different image dimensions on their website.
Section A: uses dynamic placeholder dimensions 200x200 . used image is 600x450.
Stretching is no option, as the image will not keep quality.

There is of course other way's to solve such problems, yet.. nothing seem to work stable. 
Default usage example: https://www.youtube.com/watch?v=fc-brJxcXGU
Programmatic : https://www.youtube.com/watch?v=T1rPqCRL7dw

**Comment by adityapatadia** [comment]
I think better approach should be to use dynamic image resize service like https://www.gumlet.com. There is also accompanying library for it: https://github.com/gumlet/gumlet.js

**Pull request closed without merging**